File: pvmi/proc_pid_stat_metrics.go
// proc_pid_stat_* metrics utils.

package pvmi

import (
	"bytes"
	"fmt"

	"github.com/prometheus/procfs"
)

const (
	// proc_pid_stat_state:
	PROC_PID_STAT_STATE_METRIC_NAME             = "proc_pid_stat_state"
	PROC_PID_STAT_STATE_METRIC_STATE_LABEL_NAME = "state"

	// proc_pid_stat_info:
	PROC_PID_STAT_INFO_METRIC_NAME                   = "proc_pid_stat_info"
	PROC_PID_STAT_INFO_METRIC_COMM_LABEL_NAME        = "comm"
	PROC_PID_STAT_INFO_METRIC_PPID_LABEL_NAME        = "ppid"
	PROC_PID_STAT_INFO_METRIC_PGRP_LABEL_NAME        = "pgrp"
	PROC_PID_STAT_INFO_METRIC_SESSION_LABEL_NAME     = "session"
	PROC_PID_STAT_INFO_METRIC_TTY_NR_LABEL_NAME      = "tty_nr"
	PROC_PID_STAT_INFO_METRIC_TPGID_LABEL_NAME       = "tpgid"
	PROC_PID_STAT_INFO_METRIC_FLAGS_LABEL_NAME       = "flags"
	PROC_PID_STAT_INFO_METRIC_PRIORITY_LABEL_NAME    = "priority"
	PROC_PID_STAT_INFO_METRIC_NICE_LABEL_NAME        = "nice"
	PROC_PID_STAT_INFO_METRIC_RT_PRIORITY_LABEL_NAME = "rt_priority"
	PROC_PID_STAT_INFO_METRIC_POLICY_LABEL_NAME      = "policy"
	PROC_PID_STAT_MINFLT_METRIC_NAME                 = "proc_pid_stat_minflt"

	PROC_PID_STAT_CMINFLT_METRIC_NAME = "proc_pid_stat_cminflt"

	PROC_PID_STAT_MAJFLT_METRIC_NAME = "proc_pid_stat_majflt"

	PROC_PID_STAT_CMAJFLT_METRIC_NAME = "proc_pid_stat_cmajflt"

	PROC_PID_STAT_UTIME_SECONDS_METRIC_NAME = "proc_pid_stat_utime_seconds"

	PROC_PID_STAT_STIME_SECONDS_METRIC_NAME = "proc_pid_stat_stime_seconds"

	PROC_PID_STAT_CUTIME_SECONDS_METRIC_NAME = "proc_pid_stat_cutime_seconds"

	PROC_PID_STAT_CSTIME_SECONDS_METRIC_NAME = "proc_pid_stat_cstime_seconds"

	PROC_PID_STAT_UTIME_PCT_METRIC_NAME = "proc_pid_stat_utime_pct"

	PROC_PID_STAT_STIME_PCT_METRIC_NAME = "proc_pid_stat_stime_pct"

	PROC_PID_STAT_CPU_TIME_PCT_METRIC_NAME = "proc_pid_stat_cpu_time_pct"

	PROC_PID_STAT_CCPU_TIME_PCT_METRIC_NAME = "proc_pid_stat_ccpu_time_pct"

	PROC_PID_STAT_VSIZE_METRIC_NAME = "proc_pid_stat_vsize"

	PROC_PID_STAT_RSS_METRIC_NAME = "proc_pid_stat_rss"

	PROC_PID_STAT_RSSLIM_METRIC_NAME = "proc_pid_stat_rsslim"

	PROC_PID_STAT_CPU_METRIC_NAME = "proc_pid_stat_cpu"
)

var pidStatStateMetricFmt string = fmt.Sprintf(
	`%s{%%s,%s="%%s"}`,
	PROC_PID_STAT_STATE_METRIC_NAME,
	PROC_PID_STAT_STATE_METRIC_STATE_LABEL_NAME,
)

var pidStatInfoMetricFmt string = fmt.Sprintf(
	`%s{%%s,%s="%%s",%s="%%d",%s="%%d",%s="%%d",%s="%%d",%s="%%d",%s="%%d",%s="%%d",%s="%%d",%s="%%d",%s="%%d"}`,
	PROC_PID_STAT_INFO_METRIC_NAME,
	PROC_PID_STAT_INFO_METRIC_COMM_LABEL_NAME,
	PROC_PID_STAT_INFO_METRIC_PPID_LABEL_NAME,
	PROC_PID_STAT_INFO_METRIC_PGRP_LABEL_NAME,
	PROC_PID_STAT_INFO_METRIC_SESSION_LABEL_NAME,
	PROC_PID_STAT_INFO_METRIC_TTY_NR_LABEL_NAME,
	PROC_PID_STAT_INFO_METRIC_TPGID_LABEL_NAME,
	PROC_PID_STAT_INFO_METRIC_FLAGS_LABEL_NAME,
	PROC_PID_STAT_INFO_METRIC_PRIORITY_LABEL_NAME,
	PROC_PID_STAT_INFO_METRIC_NICE_LABEL_NAME,
	PROC_PID_STAT_INFO_METRIC_RT_PRIORITY_LABEL_NAME,
	PROC_PID_STAT_INFO_METRIC_POLICY_LABEL_NAME,
)

// Invoke for new/change in procStat state:
func updateProcPidStatStateMetric(
	pmce *PidMetricsCacheEntry,
	procStat *procfs.ProcStat,
	promTs string,
	buf *bytes.Buffer,
	generatedCount *uint64,
) error {
	// Clear previous metric, if any:
	if pmce.ProcPidStatStateMetric != "" {
		buf.WriteString(pmce.ProcPidStatStateMetric + " 0 " + promTs + "\n")
		if generatedCount != nil {
			*generatedCount += 1
		}
	}
	// Update the metric:
	pmce.ProcPidStatStateMetric = fmt.Sprintf(
		pidStatStateMetricFmt,
		pmce.CommonLabels,
		procStat.State,
	)
	return nil
}

// Invoke for new/change in procStat info:
func updateProcPidStatInfoMetric(
	pmce *PidMetricsCacheEntry,
	procStat *procfs.ProcStat,
	promTs string,
	buf *bytes.Buffer,
	generatedCount *uint64,
) error {
	// Clear previous metric, if any:
	if pmce.ProcPidStatInfoMetric != "" {
		buf.WriteString(pmce.ProcPidStatInfoMetric + " 0 " + promTs + "\n")
		if generatedCount != nil {
			*generatedCount += 1
		}
	}
	// Update the metric:
	pmce.ProcPidStatInfoMetric = fmt.Sprintf(
		pidStatInfoMetricFmt,
		pmce.CommonLabels,
		procStat.Comm,
		procStat.PPID,
		procStat.PGRP,
		procStat.Session,
		procStat.TTY,
		procStat.TPGID,
		procStat.Flags,
		procStat.Priority,
		procStat.Nice,
		procStat.RTPriority,
		procStat.Policy,
	)
	return nil
}

File: tools/py/pvmi/proc_pid_stat_metrics.py
#! /usr/bin/env python3

# support for pvmi/proc_pid_metrics.go testing

import functools
from typing import List, Optional

import procfs
from metrics_common_test import TestClktckSec, TestdataProcfsRoot, TestHostname, TestJob

from .common import Metric, sanitize_label_value, ts_to_prometheus_ts


@functools.lru_cache(maxsize=None)
def make_common_labels(
    pid: int,
    tid: int = 0,
    procfs_root: str = TestdataProcfsRoot,
    _hostname: str = TestHostname,
    _job: str = TestJob,
) -> str:
    starttime = procfs.load_proc_pid_stat(
        pid, tid=tid, procfs_root=procfs_root
    ).Starttime
    if tid == 0:
        return (
            f'hostname="{_hostname}",job="{_job}",pid="{pid}",starttime="{starttime}"'
        )
    else:
        t_starttime = starttime
        if pid != tid:
            starttime = procfs.load_proc_pid_stat(
                pid, tid=0, procfs_root=procfs_root
            ).Starttime
        return (
            f'hostname="{_hostname}",job="{_job}",pid="{pid}",starttime="{starttime}"'
            f',tid="{tid}",t_starttime="{t_starttime}"'
        )


PROC_PID_STAT_STATE_METRIC_NAME = "proc_pid_stat_state"
PROC_PID_STAT_INFO_METRIC_NAME = "proc_pid_stat_info"

proc_pid_stat_categorical_metric_names = [
    PROC_PID_STAT_STATE_METRIC_NAME,
    PROC_PID_STAT_INFO_METRIC_NAME,
]


def generate_proc_pid_stat_metrics(
    proc_pid_stat: procfs.ProcStat,
    common_labels: str,
    ts: Optional[float] = None,
    prev_proc_pid_stat: Optional[procfs.ProcStat] = None,
    prev_ts: Optional[float] = None,
) -> List[Metric]:
    if ts is None:
        ts = proc_pid_stat._ts
    labels_sep = "," if common_labels else ""
    prom_ts = ts_to_prometheus_ts(ts)
    metrics = [
        Metric(
            metric=(
                f'proc_pid_stat_state{{{common_labels}{labels_sep}state="{proc_pid_stat.State}"}}'
            ),
            val=1,
            ts=prom_ts,
        ),
        Metric(
            metric=(
                f"proc_pid_stat_info{{{common_labels}{labels_sep}"
                f'comm="{sanitize_label_value(proc_pid_stat.Comm)}"'
                f',ppid="{proc_pid_stat.PPID}"'
                f',pgrp="{proc_pid_stat.PGRP}"'
                f',session="{proc_pid_stat.Session}"'
                f',tty_nr="{proc_pid_stat.TTY}"'
                f',tpgid="{proc_pid_stat.TPGID}"'
                f',flags="{proc_pid_stat.Flags}"'
                f',priority="{proc_pid_stat.Priority}"'
                f',nice="{proc_pid_stat.Nice}"'
                f',rt_priority="{proc_pid_stat.RTPriority}"'
                f',policy="{proc_pid_stat.Policy}"'
                "}"
            ),
            val=1,
            ts=prom_ts,
        ),
        Metric(
            metric=f"proc_pid_stat_minflt{{{common_labels}}}",
            val=proc_pid_stat.MinFlt,
            ts=prom_ts,
        ),
        Metric(
            metric=f"proc_pid_stat_cminflt{{{common_labels}}}",
            val=proc_pid_stat.CMinFlt,
            ts=prom_ts,
        ),
        Metric(
            metric=f"proc_pid_stat_majflt{{{common_labels}}}",
            val=proc_pid_stat.MajFlt,
            ts=prom_ts,
        ),
        Metric(
            metric=f"proc_pid_stat_cmajflt{{{common_labels}}}",
            val=proc_pid_stat.CMajFlt,
            ts=prom_ts,
        ),
        Metric(
            metric=f"proc_pid_stat_utime_seconds{{{common_labels}}}",
            val=proc_pid_stat.UTime * TestClktckSec,
            ts=prom_ts,
            valfmt=".06f",
        ),
        Metric(
            metric=f"proc_pid_stat_stime_seconds{{{common_labels}}}",
            val=proc_pid_stat.STime * TestClktckSec,
            ts=prom_ts,
            valfmt=".06f",
        ),
        Metric(
            metric=f"proc_pid_stat_cutime_seconds{{{common_labels}}}",
            val=proc_pid_stat.CUTime * TestClktckSec,
            ts=prom_ts,
            valfmt=".06f",
        ),
        Metric(
            metric=f"proc_pid_stat_cstime_seconds{{{common_labels}}}",
            val=proc_pid_stat.CSTime * TestClktckSec,
            ts=prom_ts,
            valfmt=".06f",
        ),
        Metric(
            metric=f"proc_pid_stat_vsize{{{common_labels}}}",
            val=proc_pid_stat.VSize,
            ts=prom_ts,
        ),
        Metric(
            metric=f"proc_pid_stat_rss{{{common_labels}}}",
            val=proc_pid_stat.RSS,
            ts=prom_ts,
        ),
        Metric(
            metric=f"proc_pid_stat_rsslim{{{common_labels}}}",
            val=proc_pid_stat.RSSLimit,
            ts=prom_ts,
        ),
        Metric(
            metric=f"proc_pid_stat_cpu{{{common_labels}}}",
            val=proc_pid_stat.Processor,
            ts=prom_ts,
        ),
    ]

    if prev_proc_pid_stat is not None:
        if prev_ts is None:
            prev_ts = prev_proc_pid_stat._ts
        delta_utime = proc_pid_stat.UTime - prev_proc_pid_stat.UTime
        delta_stime = proc_pid_stat.STime - prev_proc_pid_stat.STime
        delta_seconds = ts - prev_ts
        metrics.extend(
            [
                Metric(
                    metric=f"proc_pid_stat_utime_pct{{{common_labels}}}",
                    val=(delta_utime * TestClktckSec / delta_seconds) * 100.0,
                    ts=prom_ts,
                    valfmt=".02f",
                ),
                Metric(
                    metric=f"proc_pid_stat_stime_pct{{{common_labels}}}",
                    val=(delta_stime * TestClktckSec / delta_seconds) * 100.0,
                    ts=prom_ts,
                    valfmt=".02f",
                ),
                Metric(
                    metric=f"proc_pid_stat_cpu_time_pct{{{common_labels}}}",
                    val=(
                        ((delta_utime + delta_stime) * TestClktckSec / delta_seconds)
                        * 100.0
                    ),
                    ts=prom_ts,
                    valfmt=".02f",
                ),
            ]
        )

    return metrics
